Add doc for configuring plugin in Hydra 1.1 (#1506)
facebookresearch/hydra: website/docs/patterns/configuring_plugins.md

<div className="col  col--5">

```commandline title="python my_app.py -m a=1,2,3" 
[HYDRA] Launching 3 jobs locally
[HYDRA]        #0 : a=1
[HYDRA]        #1 : a=2
[HYDRA]        #2 : a=3

```
</div>
</div>

### Overriding in primary config
<ExampleGithubLink text="Example application" to="examples/patterns/configuring_plugins_primary_config"/>

In the primary config, we override `hydra.sweeper.max_batch_size`.
Running the application, we can see the the jobs are launcher in batch of one.

<div className="row">
<div className="col col--6">

```yaml title="config.yaml" 
a: 1

hydra:
  sweeper:
    max_batch_size: 1

```
</div>

<div className="col  col--6">

```commandline title="python my_app.py -m a=1,2,3" 
[HYDRA] Launching 1 jobs locally
[HYDRA]        #0 : a=1
[HYDRA] Launching 1 jobs locally
[HYDRA]        #1 : a=2
[HYDRA] Launching 1 jobs locally
[HYDRA]        #2 : a=3

```
</div>
</div>

### Extending plugin default config

<ExampleGithubLink text="Example application" to="examples/patterns/configuring_plugins_extend_defaults"/>

To achieve the same configuration, we can also create a separate config for the sweeper which helps keeping the primary config clean.

> 

Reviewer: Explain that the advantages are.
(Easier to switch between different plugin configuration, can work if the different configuration has a different schema etc).

The example may work better if we have two parameters in the plugin config.

You can show two different configurations, say - batch of 1 and batch of 2 and show how you are switching between them.

Author: > The example may work better if we have two parameters in the plugin config.

Yes - however for the examples I want to use a Plugin out of hydra-core (for simplicity and testing purpose), as a result, the choice is limited :(

Reviewer: You can talk about a hypothetical plugin without a concrete example.

For this example, we will configure and imaginary Launcher plugin called MoonLauncher.
MoonLauncher can launch Hydra applications to the Moon. It also has a simulation mode, where the application is not really launched to the moon.

...
```

Author: Thanks @omry , nice example :)

Although I still feel a runnable example is easier to comprehend than a hypothetical plugin - if you feel strongly about having the MoonLauncher I can try to update the doc further (and probably delete the new examples I've added in this PR.)

Reviewer: We will have plenty of runnable examples in all the plugins, right?
This is more about explaining the concepts.

Reviewer: This is just applying "Extending Configs" in the context of plugins. We have examples both in Extending Configs and in the plugins themselves.

Author: @omry 

Sounds good. Just so we are on the same page here: we will get rid of all the BasicSweeper examples here and only use `MoonLauncher` in this page, correct?

Reviewer: I think so. I can see the advantage of using a concrete plugin for the example. But at the same time, it makes it less useful for explaining the general concept of configuring other plugins.
Using an obviously made up plugin will probably make the point that this applies to all Hydra plugins easier to make.

<div className="row">
<div className="col col--6">

```yaml title="hydra/sweeper/custom_sweeper.yaml" 
defaults:
  - basic

> max_batch_size: 1

Reviewer: batch of 2 may make the difference more obvious. took me a moment to see the difference in output.

Author: True, I've updated the example.

Hydra plugins usually comes with sensible defaults which works with minimal configuration.
There are two primary ways to customize the configuration of a plugin:
- Overriding it directly in your primary config
- Extending the config and using it from your primary config.
	
The first method is the simplest, but it makes it harder to switch to a different plugin configuration.
The second method is a bit more complicated, but makes it easy to switch between different plugin configurations.

The following methods apply to all Hydra plugins. In the following examples, we will configure a imaginary Launcher plugin
`MoonLauncher`. The Launcher has two modes: `falcon9`, which actually launches the application to the Moon and 
`simulation` which simulates a launch.

The config schema for MoonLauncher looks like:

<div className="row">
<div className="col col--6">

```python
@dataclass
class Falcon9Conf:
  ton_fuel:  int = 10

```
</div>
<div className="col  col--6">

```python
@dataclass
class Simulation:
  ton_fuel:  int = 10
  window_size:
    width: 1024
    height: 768

```
</div>
</div>

### Overriding in primary config

> We can directly override Launcher config in primary config.

Reviewer: The problem with this approach, that you example is not exposing - is that if you want to override variables in the simulation (like window_size) in your primary config, it will not be compatible with falcon9 config anymore.
It's worth explaining this.

<div className="row">
<div className="col col--4">

```yaml title="config.yaml" 
a: 1

hydra:
  launcher:
    ton_fuel: 2

```
</div>
<div className="col col--4">

```commandline title="command-line override" 
hydra/launcher=falcon9

```
```yaml title="resulting launcher config"  {3}
hydra:
  launcher:
    ton_fuel: 2

```
</div>
<div className="col col--4">

```commandline title="command-line override" 

> hydra/launcher=simulation

Reviewer: This has a horizontal scrollbar.
Try to rename the config name for simluation to something shorter (sim?) or to reorgranize the page a bit to give this column more space.

```
```yaml title="resulting launcher config"  {3}
hydra:
  launcher:
    ton_fuel: 2
    window_size:
      width: 1024
      height: 768
```
</div>
</div>

### Extending plugin default config

This section assumes that you are familiar with the contents of [Common Patterns/Extending Configs](patterns/extending_configs.md).

Extending plugin default config has several advantages:
- Separate configuration concerns, keep primary config clean.
- Easier to switch between different plugin configurations.
- Provides flexibility when a Plugin has different modes
that requires different schema.

Say that we want to override certain values for different Launcher mode:

<div className="row">
<div className="col col--6">

```yaml title="hydra/launcher/my_falcon9" {4}
defaults:
  - falcon9

ton_fuel: 2

```
</div>
<div className="col col--6">

> ```yaml title="hydra/sweeper/my_simulation" {5}

Reviewer: make a pass on the page and make sure all the line highlights are actually focusing on something relevant.
I did not understand why you highlight line 5 here.

Reviewer: Also, make sure that the boxes are in the same size.
![image](https://user-images.githubusercontent.com/376455/116450190-db2bf900-a80f-11eb-89fe-be0265b4327b.png)

---
id: configuring_plugins
title: Configuring Plugins
---

import {ExampleGithubLink} from "@site/src/components/GithubLink"

Hydra plugins usually comes with sensible defaults which works with minimal configuration.
There are two primary ways to customize the configuration of a plugin:
- Overriding it directly in your primary config
- Extending the config and using it from your primary config.
	

> The first method is the simplest, but it makes it harder to switch to a different plugin configuration.
> The second method is a bit more complicated, but makes it easy to switch between different plugin configurations.
> 

Reviewer: ```suggestion
The first method is simpler, but it makes it harder to switch to a different plugin configuration.
The second method is a bit more complicated, but makes it easy to switch between different plugin configurations.
```

`MoonLauncher`. The Launcher has two modes: `falcon9`, which actually launches the application to the Moon and 
`sim` which simulates a launch.

The config schema for MoonLauncher looks like:

<div className="row">
<div className="col col--6">

```python
@dataclass
class Falcon9Conf:
  ton_fuel:  int = 10

```
</div>
<div className="col  col--6">

```python
@dataclass
class Sim:
  ton_fuel:  int = 10
  window_size:
    width: 1024
    height: 768

```
</div>
</div>

### Overriding in primary config
We can directly override Launcher config in primary config.

<div className="row">
<div className="col col--4">

```yaml title="config.yaml" 
a: 1

hydra:
  launcher:
    ton_fuel: 2

```
</div>
<div className="col col--4">

```commandline title="command-line override" 
hydra/launcher=falcon9

```
```yaml title="resulting launcher config"  {3}
hydra:
  launcher:
    ton_fuel: 2

```
</div>
<div className="col col--4">

```commandline title="command-line override" 
hydra/launcher=sim

```
```yaml title="resulting launcher config"  {3}
hydra:
  launcher:
    ton_fuel: 2
    window_size:
      width: 1024
      height: 768
```
</div>
</div>

> This approach works well for common configs across different modes. What if we want to override `window_size.width`  for `sim` mode? 
> Overriding the value in primary config will limit launching option to only
> `hydra/launcher=sim` - we can no longer override `hydra/launcher=falcon9` because  `falcon9` does not support configuring
> `window_size.width`. In the next section, we will look at how to extend plugin default for more configuration flexibility.

Reviewer: ```suggestion
This approach makes the assumption that the Launcher used has all the fields we are overriding.
If we wanted to override a field that exists in the Simulation Launcher but not in the Falcon9 Launcher, like window_size.width`, we would no longer be able to use the Falcon9 Launcher!  
The next section solves this problem.
```

### Extending plugin default config

This section assumes that you are familiar with the contents of [Common Patterns/Extending Configs](patterns/extending_configs.md).

Extending plugin default config has several advantages:
- Separate configuration concerns, keep primary config clean.
- Easier to switch between different plugin configurations.
- Provides flexibility when a Plugin has different modes
that requires different schema.

Say that we want to override certain values for different Launcher mode:

<div className="row">
<div className="col col--6">

> ```yaml title="hydra/launcher/my_falcon9" {4}

Reviewer: Use .yaml extention in the title for config files.
```suggestion
```yaml title="hydra/launcher/my_falcon9.yaml" {4}
```

```python
@dataclass

> class Sim:

Reviewer: We can keep the class name `Simulation`, no one said that the config group name must match the config class name exactly.
